apps/agent/app/src/script.js

```diff
@@ -47,290 +74,463 @@ const retryEvery = async (
+          tokenTransfers: [
+            {
+              amount,
+              token,
+              from: null,
+              to: event.returnValues.to,
+            },
+          ],
+        }
+      : null
+
+  const newBalances = await updateBalances(state.balances, token, settings)
+  const newTransactions = transactionDetailOptions
+    ? await updateTransactions(
+        state.transactions,
+        event,
+        transactionDetailOptions
+      )
+    : state.transactions
+
+  return {
+    ...state,
+    balances: newBalances,
+    transactions: newTransactions,
+  }
+}
+
+async function newExecution(state, event, settings) {
+  const {
+    transactionHash,
+    returnValues: { ethValue },
+  } = event
+  // Let's try to find some more information about this particular execution
+  // by using the transaction receipt:
+  //   - To address
+  //   - Any internal token transfers to / from this agent app
+  const transactionReceipt = await app
+    .web3Eth('getTransactionReceipt', transactionHash)
+    .toPromise()
+  const tokenTransfers = findTransfersFromReceipt(transactionReceipt)
+    .map(({ token, returnData }) => {
+      const { from, to, value } = returnData
+      const fromAgent = addressesEqual(
+        from.toLowerCase(),
+        settings.proxyAddress
+      )
+      const toAgent = addressesEqual(to.toLowerCase(), settings.proxyAddress)
+      // Filter out token transfers not to the agent
+      return fromAgent || toAgent
+        ? {
+            token,
+            amount: value,
+            from: fromAgent ? null : from,
+            to: toAgent ? null : to,
+          }
+        : null
+    })
+    .filter(Boolean)
+
+  if (ethValue && ethValue !== '0') {
+    tokenTransfers.push({
+      amount: ethValue,
+      from: null,
+      to: transactionReceipt.to,
+      token: settings.ethToken.address,
+    })
+  }
+  // Also try to find the target contract for contract interactions,
+  // which do not necessarily have token transfers
+  const targetContract = findTargetFromReceipt(
+    transactionReceipt,
+    settings.proxyAddress
+  )
+  let newBalances = state.balances
+  const updatedTokenAddresses = tokenTransfers.map(({ token }) => token)
+  for (const address of updatedTokenAddresses) {
+    newBalances = await updateBalances(newBalances, address, settings)
+  }
+
+  const newTransactions = await updateTransactions(state.transactions, event, {
+    // TODO: fetch description via radspec / on-chain registry fallback
+    // (see https://metamask.github.io/metamask-docs/Best_Practices/Registering_Function_Names)
+    description: 'Contract interaction',
+    tokenTransfers,
+    targetContract,
+  })
+  return {
+    ...state,
+    balances: newBalances,
+    transactions: newTransactions,
+  }
+}
+
+/***********************
+ *                     *
+ *    Token Helpers    *
+ *                     *
+ ***********************/
+
+async function loadTokenBalances(state, includedTokenAddresses, settings) {
```

Comment:
Let's create this section of the script in Finance as well, and make all the relevant changes so it looks like this :).

Also here, let's rename `includedTokenAddresses` to just `tokenAddresses`.

Reply:
[Made an issue to keep track of this!](https://github.com/aragon/aragon-apps/issues/1091)